File: forwindows/src/AnyClipCore/UpdateChecker.cs
// forwindows/src/AnyClipCore/UpdateChecker.cs
using System.Text.Json;

namespace AnyClip.Core;

public abstract record UpdateStatus
{
    public sealed record UpToDate(string Current) : UpdateStatus;
    public sealed record Available(string Latest, string Url) : UpdateStatus;
    public sealed record Failed(string Reason) : UpdateStatus;
}

/// Pure update detection. Network IO is injected via `fetch` so the
/// parse/compare logic is unit-testable without hitting GitHub.
public static class UpdateChecker
{
    public const string ReleasesApiUrl =
        "https://api.github.com/repos/gum798/AnyClip/releases/latest";
    public const string ReleasesPageUrl =
        "https://github.com/gum798/AnyClip/releases/latest";

    public static string? ParseLatestTag(string json)
    {
        try
        {
            using var doc = JsonDocument.Parse(json);
            if (doc.RootElement.TryGetProperty("tag_name", out var t)
                && t.ValueKind == JsonValueKind.String
                && t.GetString() is { } tag)
            {
                var v = tag.StartsWith('v') ? tag[1..] : tag;
                return string.IsNullOrEmpty(v) ? null : v;
            }
        }
        catch (JsonException) { /* fall through */ }
        return null;
    }

    /// Semver-ish: numeric components dominate; a pre-release ("-" suffix)
    /// ranks below the same core; non-numeric components sort low. Returns
    /// negative / 0 / positive like a comparator.
    public static int CompareVersions(string a, string b)
    {
        var (na, preA) = Parse(a);
        var (nb, preB) = Parse(b);
        int n = Math.Max(na.Count, nb.Count);
        for (int i = 0; i < n; i++)
        {
            int x = i < na.Count ? na[i] : 0;
            int y = i < nb.Count ? nb[i] : 0;
            if (x != y) return x < y ? -1 : 1;
        }
        if (preA != preB) return preA ? -1 : 1;
        return 0;
    }

    private static (List<int> nums, bool isPre) Parse(string v)
    {
        var core = v.Split('-', 2)[0];
        var nums = core.Split('.')
            .Select(s => int.TryParse(s, out var n) ? n : -1).ToList();
        return (nums, v.Contains('-'));
    }

    public static async Task<UpdateStatus> CheckForUpdateAsync(
        string current, Func<Task<string>> fetch)
    {
        string body;
        try { body = await fetch(); }
        catch (Exception e) { return new UpdateStatus.Failed(e.Message); }
        var latest = ParseLatestTag(body);
        if (latest is null) return new UpdateStatus.Failed("could not parse latest release");
        return CompareVersions(current, latest) < 0
            ? new UpdateStatus.Available(latest, ReleasesPageUrl)
            : new UpdateStatus.UpToDate(current);
    }
}

File: formacOS/Sources/AnyClipCore/UpdateChecker.swift
import Foundation

public enum UpdateStatus: Equatable, Sendable {
    case upToDate(current: String)
    case available(latest: String, url: String)
    case failed(reason: String)
}

/// Pure update detection. Network IO is injected via `fetch`, so the
/// parse/compare logic is unit-testable without hitting GitHub.
public enum UpdateChecker {
    public static let releasesApiURL =
        "https://api.github.com/repos/gum798/AnyClip/releases/latest"
    public static let releasesPageURL =
        "https://github.com/gum798/AnyClip/releases/latest"

    /// `tag_name` from GitHub releases JSON, leading "v" stripped. nil if absent/malformed.
    public static func parseLatestTag(_ json: String) -> String? {
        guard let data = json.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let tag = obj["tag_name"] as? String else { return nil }
        let v = tag.hasPrefix("v") ? String(tag.dropFirst()) : tag
        return v.isEmpty ? nil : v
    }

    /// Semver-ish: numeric components dominate; a pre-release ("-" suffix)
    /// ranks below the same core; non-numeric components sort low.
    public static func compareVersions(_ a: String, _ b: String) -> ComparisonResult {
        let pa = parse(a), pb = parse(b)
        let n = max(pa.nums.count, pb.nums.count)
        for i in 0..<n {
            let x = i < pa.nums.count ? pa.nums[i] : 0
            let y = i < pb.nums.count ? pb.nums[i] : 0
            if x != y { return x < y ? .orderedAscending : .orderedDescending }
        }
        if pa.isPre != pb.isPre { return pa.isPre ? .orderedAscending : .orderedDescending }
        return .orderedSame
    }

    private static func parse(_ v: String) -> (nums: [Int], isPre: Bool) {
        let core = v.split(separator: "-", maxSplits: 1).first.map(String.init) ?? v
        let nums = core.split(separator: ".").map { Int($0) ?? -1 }
        return (nums, v.contains("-"))
    }

    public static func checkForUpdate(
        current: String, fetch: () async throws -> String
    ) async -> UpdateStatus {
        let body: String
        do { body = try await fetch() }
        catch { return .failed(reason: "\(error)") }
        guard let latest = parseLatestTag(body) else {
            return .failed(reason: "could not parse latest release")
        }
        return compareVersions(current, latest) == .orderedAscending
            ? .available(latest: latest, url: releasesPageURL)
            : .upToDate(current: current)
    }
}
